src/sessions.py

```diff
@@ -0,0 +1,142 @@
+        '''Check if the previous session ended with a failed command and
+        suggest sending the session'''
+        cursor = self._sql('''
+            SELECT COUNT(*) FROM bayesdb_session_entries
+                WHERE completed=0 AND session_id=?;
+        ''', (self.session_id-1,))
+        uncompleted_entries = int(cursor.next()[0])
+        if uncompleted_entries > 0:
+            self._print('WARNING: Previous session contains uncompleted entries. ' +
+                    'This may be due to a bad termination or crash of the ' +
+                    'previous session. Consider uploading the session with send_session_data().')
+        return uncompleted_entries
+
+    def _bql_trace(self, query, bindings):
+        return self._trace("bql", query, bindings)
+
+    def _sql_trace(self, query, bindings):
+        return self._trace("sql", query, bindings)
+
+    def clear_all_sessions(self):
+        self._sql('DELETE FROM bayesdb_session_entries;')
+        self._sql('DELETE FROM bayesdb_session;')
+        self._sql('''
+            DELETE FROM sqlite_sequence
+                WHERE name="bayesdb_session"
+                OR name="bayesdb_session_entries";
+        ''')
+        self._start_new_session()
+
+    def list_sessions(self):
+        """Lists all saved sessions with the number of entries in each, and
+        whether they were sent or not."""
+        return self._sql('SELECT * FROM bayesdb_session;')
+    
+    def current_session_id(self):
+        """Returns the current integer session id."""
+        return self.session_id
+    
+    def dump_session_as_json(self, session_id):
+        """Returns a JSON string representing the list of SQL or BQL entries
+        (e.g.  queries) executed within session `session_id`."""
+        if session_id > self.session_id or session_id < 1:
+            raise ValueError('No such session (%d)' % session_id)
+        entries = self._sql('''
+            SELECT * FROM bayesdb_session_entries
+                WHERE session_id == ?
+                ORDER BY time DESC''', (session_id,))
+        return json.dumps(list(entries))
+    
+    def dump_current_session_as_json(self):
+        """Returns a JSON string representing the current sesion (see
+        `dump_session_as_json`)"""
+        return self.dump_session_as_json(self.session_id)
+    
+    def send_session_data(self):
+        """Send all saved session history."""
+        probcomp_url = 'http://probcomp.csail.mit.edu/bayesdb/save_sessions.cgi'
+        for id in range(1, self.session_id+1):
+            self._print('Sending session %d to %s ...' % (id, probcomp_url))
+            json_string = self.dump_session_as_json(id)
+            self._print(json_string)
+            r = requests.post(probcomp_url,
+                    data={'session_json' : json_string})
+            self._print('Response: %s' % (r.text,))
+
+    def start_saving_sessions(self):
+        self.bdb.trace(self.bql_tracer)
```

Comment:
Is there a way to check if there are other session saver traces lingering? If you just always untrace before tracing, does that do the right thing? What if the lambdas are different?